### PR TITLE
added multi-curl

### PR DESCRIPTION
**Pull request by paulhandy**

### CODE REVIEW

File: curl/src/curl.rs [hunk 1]
```diff
@@ -1,14 +1,69 @@
-use trytes::Trit;
+use constants::*;
+//use trytes::Trit;
```
Comment (th0br0) [inline curl/src/curl.rs]: Remove comment.

File: curl/src/curl.rs [hunk 2]
```diff
@@ -1,14 +1,69 @@
...
-pub trait Curl {
-    /// Absorb a `&[Trit]` into the sponge
-    fn absorb(&mut self, trits: &[Trit]);
-    /// Squeeze the sponge and return a `Trinary` with `tritCount` trits
-    fn squeeze(&mut self, trit_count: usize) -> Vec<Trit>;
-    /// Transforms the sponge
+
+pub trait Sponge {
     fn transform(&mut self);
-    /// Resets the sponge's internal state.
```
Comment (th0br0) [inline curl/src/curl.rs]: Please add the comments back in :'(

File: curl/src/curl.rs [hunk 3]
```diff
@@ -1,14 +1,69 @@
...
+        *self
+    }
+}
+
+
+impl<T> Curl<T>
+    where T: Clone + Copy + Sized,
+          Curl<T>: Sponge
+{
+    fn absorb(&mut self, trits: &[T], offset_in: usize, length_in: usize) {
```
Comment (th0br0) [inline curl/src/curl.rs]: Where have all my comments gone

File: curl/src/lib.rs [hunk 4]
```diff
@@ -1,8 +1,12 @@
+extern crate rayon;
 extern crate iota_trytes as trytes;
 
+mod indices;
 pub mod constants;
 pub mod curl;
+//pub mod scurl;
```
Comment (th0br0) [inline curl/src/lib.rs]: remove comment

File: curl/src/pair.rs [hunk 5]
```diff
@@ -0,0 +1,58 @@
+use rayon::prelude::*;
+use constants::*;
+use trytes::*;
+use curl::*;
+use indices::*;
+//use std::mem;
```
Comment (th0br0) [inline curl/src/pair.rs]: remove

File: curl/src/pair.rs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+/// Tuple implementation of the `Sponge` for Curl
+impl Sponge for Curl<BCTrit> {
+    fn transform(&mut self) {
+        // Required memory space type for computation
+        //let mut scratchpad: Vec<BCTrit> = self.state.clone();
+        let STATE_INDICES: Vec<_> = (0..STATE_LENGTH).collect();
+        let mut scratchpad: Vec<BCTrit> = self.state.iter().map(|&c| (c.0, c.1)).collect();
+
+        for _ in 0..NUMBER_OF_ROUNDS {
+            scratchpad = STATE_INDICES
```
Comment (th0br0) [inline curl/src/pair.rs]: `scratchpad = (0..STATE_LENGTH).into_par_iter()`

File: curl/src/simple.rs [hunk 7]
```diff
@@ -2,41 +2,19 @@ use constants::*;
...
-        type Space = i8;
-
-        let mut scratchpad: [Space; STATE_LENGTH] = [0; STATE_LENGTH];
+        let mut scratchpad: [Trit; STATE_LENGTH] = [0; STATE_LENGTH];
         let mut scratchpad_index: usize = 0;
         let mut scratchpad_index_save: usize;
 
         for _ in 0..NUMBER_OF_ROUNDS {
-            let state_space: Vec<Space> = self.state.iter().map(|&c| c as Space).collect();
+            //let state_space: Vec<Space> = self.state.iter().map(|&c| c as Space).collect();
```
Comment (th0br0) [inline curl/src/simple.rs]: remove

File: curl/src/simple.rs [hunk 8]
```diff
@@ -2,41 +2,19 @@ use constants::*;
...
+            let state_space: Vec<Trit> = self.state.iter().cloned().collect();
+            //scratchpad.clone_from_slice(state_space.as_slice());
```
Comment (th0br0) [inline curl/src/simple.rs]: remove

File: curl/src/tests.rs [hunk 9]
```diff
@@ -51,15 +53,15 @@ pub mod testsuite {
                               9999999999999999999999999999999999999999999999999999999999999\
                               999999999999999999999999999999T999999999999999999999999999999\
                               99999999999999999999999OLOB99999999999999999999999"
-                .chars()
-                .collect();
+                    .chars()
```
Comment (th0br0) [inline curl/src/tests.rs]: Did you run `rustfmt`?

File: curl/src/curl.rs (hunk 3 above)
Comment (th0br0) [reply curl/src/curl.rs]: Also, I'm not happy with specifying `offset_in` and `length_in`.
That's what slices are for, they already carry the length around.

File: curl/src/tests.rs [hunk 10]
```diff
@@ -51,15 +53,15 @@ pub mod testsuite {
...
 
             let ex_hash: Trinary = "TAQCQAEBHLLYKAZWMNSXUPWQICMFSKWPEGQBNM9AQMGLFZGME9REOZTQIJQRKYH\
                              DANIYSMFYPVABX9999"
-                .chars()
-                .collect();
+                    .chars()
+                    .collect();
 
-            curl.absorb(&trans.trits());
+            curl.absorb(&trans.trits(), 0, trans.trits().len());
```
Comment (th0br0) [inline curl/src/tests.rs]: Referring to the offset/length thing, you'd just do
```rust
let trits = trans.trits();
curl.absorb(&trits); 
```

here

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: curl/src/curl.rs (hunk 3 above)
Comment (paulhandy) [reply curl/src/curl.rs]: I suppose that's fair enough.

Comment (paulhandy) [reply curl/src/curl.rs]: Also, sorry about your comments. I'll put them back

File: curl/src/pair.rs (hunk 6 above)
Comment (paulhandy) [reply curl/src/pair.rs]: that does not compile

File: curl/src/tests.rs (hunk 9 above)
Comment (paulhandy) [reply curl/src/tests.rs]: vim-racer

**Review by th0br0 — APPROVED** [review]

**Pull request merged**